File: lib/java/src/main/java/uapi/internal/binary/EncodeKeys.java
package uapi.internal.binary;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class EncodeKeys {
    static Object encodeKeys(Object given, BinaryEncoding binaryEncoder) {
        if (given == null) {
            return given;
        } else if (given instanceof final Map<?, ?> m) {
            final var newMap = new HashMap<Object, Object>();

            for (final var e : m.entrySet()) {
                final var key = e.getKey();

                final Object finalKey;
                if (binaryEncoder.encodeMap.containsKey(key)) {
                    finalKey = binaryEncoder.encodeMap.get(key);
                } else {
                    finalKey = key;
                }

                final var encodedValue = encodeKeys(e.getValue(), binaryEncoder);

                newMap.put(finalKey, encodedValue);
            }

            return newMap;
        } else if (given instanceof List<?> l) {
            return l.stream().map(e -> encodeKeys(e, binaryEncoder)).toList();
        } else {
            return given;
        }
    }
}

File: lib/ts/uapi/internal/binary/EncodeKeys.ts
import { BinaryEncoding } from 'uapi/internal/binary/BinaryEncoding';

export function encodeKeys(given: any, binaryEncoder: BinaryEncoding): any {
    if (given === null || given === undefined) {
        return given;
    } else if (typeof given === 'object' && !Array.isArray(given)) {
        const newMap = new Map<any, any>();

        for (const [key, value] of Object.entries(given)) {
            const finalKey = binaryEncoder.encodeMap.has(key) ? binaryEncoder.encodeMap.get(key) : key;
            const encodedValue = encodeKeys(value, binaryEncoder);

            newMap.set(finalKey, encodedValue);
        }

        return newMap;
    } else if (Array.isArray(given)) {
        return given.map((value) => encodeKeys(value, binaryEncoder));
    } else {
        return given;
    }
}
